agent/rpc_from.go

```diff
@@ -126,8 +141,46 @@ func (a *orbAgent) handleAgentStop(payload fleet.AgentStopRPCPayload) {
 	panic(fmt.Sprintf("control plane requested we terminate, reason: %s", payload.Reason))
 }
 
+func contains(s []string, e string) (bool, int) {
+	for i, a := range s {
+		if a == e {
+			return true, i
+		}
+	}
+	return false, 0
+}
+
 func (a *orbAgent) handleAgentGroupRemoval(rpc fleet.GroupRemovedRPCPayload) {
 	a.unsubscribeGroupChannel(rpc.ChannelID)
+
+	groupRemoved := a.groupsInfos[rpc.AgentGroupID]
+
+	for _, groupInfo := range a.groupsInfos {
+		if !reflect.DeepEqual(groupInfo, groupRemoved) {
+			for _, groupPolicy := range groupInfo.policyIDs {
+				if removePolicy, index := contains(groupRemoved.policyIDs, groupPolicy); removePolicy {
+					groupRemoved.policyIDs[index] = groupRemoved.policyIDs[len(groupRemoved.policyIDs)-1]
+					groupRemoved.policyIDs[len(groupRemoved.policyIDs)-1] = ""
+					groupRemoved.policyIDs = groupRemoved.policyIDs[:len(groupRemoved.policyIDs)-1]
+				}
+			}
+		}
+	}
+
+	for _, policy := range groupRemoved.policyIDs {
```

Comment:
So I think this logic is not right because their policy can "belong" to multiple groups. I think that in addition to storing the list of policy IDs per group, we also need to store the list of group IDs per policy, and only remove the policy if it's not being used by any more groups (otherwise it just removes that group id from the list)